scripts/2.TCGA-MLexample_Pathway.py

```diff
@@ -0,0 +1,535 @@
+y.value_counts(True)
+
+
+# In[28]:
+
+y.head(5)
+
+
+# ## Set aside 10% of the data for testing
+
+# In[29]:
+
+# Make sure the splits have equal tissue partitions
+X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, random_state=0,
+                                                    stratify=clinical['disease'])
+'Size: {:,} features, {:,} training samples, {:,} testing samples'.format(len(X.columns),
+                                                                          len(X_train), len(X_test))
+
+
+# ## Median absolute deviation feature selection
+
+# In[30]:
+
+def fs_mad(x, y):
+    """    
+    Get the median absolute deviation (MAD) for each column of x
+    """
+    scores = mad(x) 
+    return scores, np.array([np.NaN]*len(scores))
+
+# select the top features with the highest MAD
+feature_select = SelectKBest(fs_mad, k=n_feature_kept)
+
+
+# ## Define pipeline and Cross validation model fitting
+
+# In[31]:
+
+# Include loss='log' in param_grid doesn't work with pipeline somehow
+clf = SGDClassifier(random_state=0, class_weight='balanced',
+                    loss=param_fixed['loss'], penalty=param_fixed['penalty'])
+
+# joblib is used to cross-validate in parallel by setting `n_jobs=-1` in GridSearchCV
+# Supress joblib warning. See https://github.com/scikit-learn/scikit-learn/issues/6370
+warnings.filterwarnings('ignore', message='Changing the shape of non-C contiguous array')
+clf_grid = grid_search.GridSearchCV(estimator=clf, param_grid=param_grid, n_jobs=-1, scoring='roc_auc')
+pipeline = make_pipeline(
+    feature_select,  # Feature selection
+    StandardScaler(),  # Feature scaling
+    clf_grid)
+
+
+# In[32]:
+
+get_ipython().run_cell_magic('time', '', '# Fit the model (the computationally intensive part)\npipeline.fit(X=X_train, y=y_train)\nbest_clf = clf_grid.best_estimator_\nfeature_mask = feature_select.get_support()  # Get a boolean array indicating the selected features')
+
+
+# In[33]:
+
+clf_grid.best_params_
+
+
+# In[34]:
+
+best_clf
+
+
+# ## Visualize hyperparameters performance
+
+# In[35]:
+
+def grid_scores_to_df(grid_scores):
+    """
+    Convert a sklearn.grid_search.GridSearchCV.grid_scores_ attribute to 
+    a tidy pandas DataFrame where each row is a hyperparameter-fold combinatination.
+    """
+    rows = list()
+    for grid_score in grid_scores:
+        for fold, score in enumerate(grid_score.cv_validation_scores):
+            row = grid_score.parameters.copy()
+            row['fold'] = fold
+            row['score'] = score
+            rows.append(row)
+    df = pd.DataFrame(rows)
+    return df
+
+
+# ## Process Mutation Matrix
+
+# In[36]:
+
+cv_score_df = grid_scores_to_df(clf_grid.grid_scores_)
+cv_score_df.head(2)
+
+
+# In[37]:
+
+# Cross-validated performance distribution
+facet_grid = sns.factorplot(x='l1_ratio', y='score', col='alpha',
+    data=cv_score_df, kind='violin', size=4, aspect=1)
+facet_grid.set_ylabels('AUROC');
+
+
+# In[38]:
```

Comment:
oh, I see what you're saying...we are not comprehensive in hyperparameter search space... we can explore more of this space in the cognoma implementation